File: paddle/fluid/framework/attribute.cc
/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/fluid/framework/attribute.h"
#include "paddle/phi/common/complex.h"
#include "paddle/phi/common/scalar.h"
#include "paddle/utils/blank.h"

namespace paddle::framework {

paddle::any GetAttrValue(const Attribute& attr) {
  switch (AttrTypeID(attr)) {
    case proto::AttrType::INT:
      return PADDLE_GET_CONST(int, attr);
    case proto::AttrType::FLOAT:
      return PADDLE_GET_CONST(float, attr);
    case proto::AttrType::FLOAT64:
      return PADDLE_GET_CONST(double, attr);
    case proto::AttrType::STRING:
      return PADDLE_GET_CONST(std::string, attr);
    case proto::AttrType::INTS:
      return PADDLE_GET_CONST(std::vector<int>, attr);
    case proto::AttrType::FLOATS:
      return PADDLE_GET_CONST(std::vector<float>, attr);
    case proto::AttrType::STRINGS:
      return PADDLE_GET_CONST(std::vector<std::string>, attr);
    case proto::AttrType::BOOLEAN:
      return PADDLE_GET_CONST(bool, attr);
    case proto::AttrType::BOOLEANS:
      return PADDLE_GET_CONST(std::vector<bool>, attr);
    case proto::AttrType::LONG:
      return PADDLE_GET_CONST(int64_t, attr);
    case proto::AttrType::LONGS:
      return PADDLE_GET_CONST(std::vector<int64_t>, attr);
    case proto::AttrType::FLOAT64S:
      return PADDLE_GET_CONST(std::vector<double>, attr);
    case proto::AttrType::VAR:
      return PADDLE_GET_CONST(VarDesc*, attr);
    case proto::AttrType::VARS:
      return PADDLE_GET_CONST(std::vector<VarDesc*>, attr);
    case proto::AttrType::BLOCK:
      return PADDLE_GET_CONST(BlockDesc*, attr);
    case proto::AttrType::BLOCKS:
      return PADDLE_GET_CONST(std::vector<BlockDesc*>, attr);
    case proto::AttrType::SCALAR:
      return PADDLE_GET_CONST(paddle::experimental::Scalar, attr);
    case proto::AttrType::SCALARS:
      return PADDLE_GET_CONST(std::vector<paddle::experimental::Scalar>, attr);
    default:
      PADDLE_THROW(common::errors::Unimplemented(
          "Unsupported Attribute value type `%s` for phi.",
          common::demangle(attr.type().name())));
  }
}

Attribute GetAttrValue(const proto::OpDesc::Attr& attr_desc) {
  switch (attr_desc.type()) {
    case proto::AttrType::BOOLEAN: {
      return attr_desc.b();
    }
    case proto::AttrType::INT: {
      return attr_desc.i();
    }
    case proto::AttrType::FLOAT: {
      return attr_desc.f();
    }
    case proto::AttrType::STRING: {
      return attr_desc.s();
    }
    case proto::AttrType::FLOAT64: {
      return attr_desc.float64();
    }
    case proto::AttrType::BOOLEANS: {
      std::vector<bool> val(attr_desc.bools_size());
      for (int i = 0; i < attr_desc.bools_size(); ++i) {
        val[i] = attr_desc.bools(i);
      }
      return val;
    }
    case proto::AttrType::INTS: {
      std::vector<int> val(attr_desc.ints_size());
      for (int i = 0; i < attr_desc.ints_size(); ++i) {
        val[i] = attr_desc.ints(i);
      }
      return val;
    }
    case proto::AttrType::FLOATS: {
      std::vector<float> val(attr_desc.floats_size());
      for (int i = 0; i < attr_desc.floats_size(); ++i) {
        val[i] = attr_desc.floats(i);
      }
      return val;
    }
    case proto::AttrType::STRINGS: {
      std::vector<std::string> val(attr_desc.strings_size());
      for (int i = 0; i < attr_desc.strings_size(); ++i) {
        val[i] = attr_desc.strings(i);
      }
      return val;
    }
    case proto::AttrType::LONG: {
      return attr_desc.l();
    }
    case proto::AttrType::LONGS: {
      std::vector<int64_t> val(attr_desc.longs_size());
      for (int i = 0; i < attr_desc.longs_size(); ++i) {
        val[i] = attr_desc.longs(i);
      }
      return val;
    }

    case proto::AttrType::FLOAT64S: {
      std::vector<double> val(attr_desc.float64s_size());
      for (int i = 0; i < attr_desc.float64s_size(); ++i) {
        val[i] = attr_desc.float64s(i);
      }
      return val;
    }

    case proto::AttrType::SCALAR: {
      return MakeScalarFromProto(attr_desc.scalar());
    }

    case proto::AttrType::SCALARS: {
      std::vector<paddle::experimental::Scalar> val(attr_desc.scalars_size());
      for (int i = 0; i < attr_desc.scalars_size(); ++i) {
        val[i] = MakeScalarFromProto(attr_desc.scalars(i));
      }
      return val;
    }

    default:
      PADDLE_THROW(common::errors::Unavailable("Unsupported attribute type %d.",
                                               attr_desc.type()));
  }
  return paddle::blank();
}

Attribute GetAttrValue(const proto::VarDesc::Attr& attr_desc) {
  switch (attr_desc.type()) {
    case proto::AttrType::INT: {
      return attr_desc.i();
    }
    case proto::AttrType::STRING: {
      return attr_desc.s();
    }
    case proto::AttrType::INTS: {
      std::vector<int> val(attr_desc.ints_size());
      for (int i = 0; i < attr_desc.ints_size(); ++i) {
        val[i] = attr_desc.ints(i);
      }
      return val;
    }
    default:
      PADDLE_THROW(common::errors::Unavailable("Unsupported attribute type %d.",
                                               attr_desc.type()));
  }
  return paddle::blank();
}

paddle::experimental::Scalar MakeScalarFromProto(const proto::Scalar& v) {
  auto data_type = v.type();
  switch (data_type) {
    case proto::Scalar_Type_BOOLEAN:
      return paddle::experimental::Scalar(v.b());
    case proto::Scalar_Type_LONG:
      return paddle::experimental::Scalar(v.i());
    case proto::Scalar_Type_FLOAT64:
      return paddle::experimental::Scalar(v.r());
    case proto::Scalar_Type_COMPLEX128: {
      phi::dtype::complex<double> value(v.c().r(), v.c().i());
      return paddle::experimental::Scalar(value);
    }
    default:
      PADDLE_THROW(common::errors::InvalidArgument(
          "Expected scalar of type boolean, "
          "integer, floating point or complex."));
      break;
  }
  return paddle::experimental::Scalar();
}

proto::Scalar MakeScalarProto(const paddle::experimental::Scalar& v) {
  proto::Scalar s;
  auto data_type = v.dtype();
  switch (data_type) {
    case phi::DataType::BOOL:
      s.set_b(v.to<bool>());
      s.set_type(proto::Scalar_Type_BOOLEAN);
      break;
    case phi::DataType::INT8:
    case phi::DataType::UINT8:
    case phi::DataType::INT16:
    case phi::DataType::UINT16:
    case phi::DataType::INT32:
    case phi::DataType::UINT32:
    case phi::DataType::INT64:
    case phi::DataType::UINT64:
      s.set_i(v.to<int64_t>());
      s.set_type(proto::Scalar_Type_LONG);
      break;
    case phi::DataType::FLOAT16:
    case phi::DataType::BFLOAT16:
    case phi::DataType::FLOAT32:
    case phi::DataType::FLOAT64:
      s.set_r(v.to<double>());
      s.set_type(proto::Scalar_Type_FLOAT64);
      break;
    case phi::DataType::COMPLEX64:
    case phi::DataType::COMPLEX128: {
      auto value = v.to<phi::dtype::complex<double>>();
      auto* complex = s.mutable_c();
      complex->set_r(value.real);
      complex->set_i(value.imag);
      s.set_type(proto::Scalar_Type_COMPLEX128);
      break;
    }
    case phi::DataType::UNDEFINED:
    case phi::DataType::PSTRING:
    case phi::DataType::NUM_DATA_TYPES:
      PADDLE_THROW(common::errors::InvalidArgument(
          "Expected scalar of type boolean, "
          "integer, floating point or complex."));
      break;
    default:
      PADDLE_THROW(common::errors::InvalidArgument(
          "Expected scalar of type boolean, "
          "integer, floating point or complex."));
      break;
  }
  return s;
}

paddle::experimental::Scalar MakeScalarFromAttribute(const Attribute& v) {
  auto attr_type = static_cast<proto::AttrType>(v.index() - 1);
  switch (attr_type) {
    case proto::AttrType::SCALAR:
      return paddle::experimental::Scalar(
          PADDLE_GET_CONST(paddle::experimental::Scalar, v));
    case proto::AttrType::BOOLEAN:
      return paddle::experimental::Scalar(PADDLE_GET_CONST(bool, v));
    case proto::AttrType::INT:
      return paddle::experimental::Scalar(PADDLE_GET_CONST(int, v));
    case proto::AttrType::LONG:
      return paddle::experimental::Scalar(PADDLE_GET_CONST(int64_t, v));
    case proto::AttrType::FLOAT:
      return paddle::experimental::Scalar(PADDLE_GET_CONST(float, v));
    case proto::AttrType::FLOAT64:
      return paddle::experimental::Scalar(PADDLE_GET_CONST(double, v));
    default:
      PADDLE_THROW(common::errors::InvalidArgument(
          "Unable to construct Scalar from given Attribute of type %s",
          attr_type));
  }
}

std::vector<paddle::experimental::Scalar> MakeScalarsFromAttribute(
    const Attribute& v) {
  auto attr_type = static_cast<proto::AttrType>(v.index() - 1);
  switch (attr_type) {
    case proto::AttrType::SCALARS:
      return PADDLE_GET_CONST(std::vector<paddle::experimental::Scalar>, v);
    case proto::AttrType::BOOLEANS:
      return experimental::WrapAsScalars(
          PADDLE_GET_CONST(std::vector<bool>, v));
    case proto::AttrType::INTS:
      return experimental::WrapAsScalars(PADDLE_GET_CONST(std::vector<int>, v));
    case proto::AttrType::LONGS:
      return experimental::WrapAsScalars(
          PADDLE_GET_CONST(std::vector<int64_t>, v));
    case proto::AttrType::FLOATS:
      return experimental::WrapAsScalars(
          PADDLE_GET_CONST(std::vector<float>, v));
    case proto::AttrType::FLOAT64S:
      return experimental::WrapAsScalars(
          PADDLE_GET_CONST(std::vector<double>, v));
    default:
      PADDLE_THROW(common::errors::InvalidArgument(
          "Unable to construct Scalars from given Attribute of type %s",
          attr_type));
  }
}

void CanonicalizeScalarAttrs(const proto::OpProto& op_proto,
                             AttributeMap* attrs) {
  PADDLE_ENFORCE_NOT_NULL(
      attrs, common::errors::InvalidArgument("attrs can not be nullptr"));
  for (auto& attr : op_proto.attrs()) {
    proto::AttrType attr_type = attr.type();
    const std::string& attr_name = attr.name();
    auto it = attrs->find(attr_name);
    if (it == attrs->end()) {
      continue;
    }
    proto::AttrType actual_attr_type = AttrTypeID(it->second);
    if (actual_attr_type == attr_type) {
      continue;
    }
    if (actual_attr_type == proto::AttrType::VAR ||
        actual_attr_type == proto::AttrType::VARS) {
      continue;  // VAR& VARS are not proper attribute
    }
    if (attr_type == proto::AttrType::SCALAR) {
      it->second = Attribute(MakeScalarFromAttribute(it->second));
    } else if (attr_type == proto::AttrType::SCALARS) {
      it->second = Attribute(MakeScalarsFromAttribute(it->second));
    }
  }
}
}  // namespace paddle::framework

File: python/paddle/tensor/attribute.py
#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# TODO: define functions to get tensor attributes

import numpy as np

import paddle
from paddle import _C_ops

from ..base.data_feeder import check_type, check_variable_and_dtype
from ..base.framework import in_dynamic_or_pir_mode, use_pir_api
from ..common_ops_import import Variable
from ..framework import LayerHelper, core
from .creation import _complex_to_real_dtype, assign

__all__ = []


def rank(input):
    """

    Returns the number of dimensions for a tensor, which is a 0-D int32 Tensor.

    Args:
        input (Tensor): The input Tensor with shape of :math:`[N_1, N_2, ..., N_k]`, the data type is arbitrary.

    Returns:
        Tensor, the output data type is int32.: The 0-D tensor with the dimensions of the input Tensor.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> input = paddle.rand((3, 100, 100))
            >>> rank = paddle.rank(input)
            >>> print(rank.numpy())
            3
    """
    check_type(input, 'input', (Variable, paddle.pir.Value), 'input')
    ndims = len(input.shape)
    out = assign(np.array(ndims, 'int32'))

    return out


def shape(input):
    """
    Get the shape of the input.

    .. code-block:: text

        Case1:
            Given N-D Tensor:
                input = [ [1, 2, 3, 4], [5, 6, 7, 8] ]

            Then:
                input.shape = [2, 4]

        Case2:
            Given SelectedRows:
                input.rows = [0, 4, 19]
                input.height = 20
                input.value = [ [1, 2], [3, 4], [5, 6] ]  # inner tensor
            Then:
                input.shape = [3, 2]

    Args:
        input (Variable): The input can be N-D Tensor or SelectedRows with data type bool, bfloat16, float16, float32, float64, int32, int64.
                          If input variable is type of SelectedRows, returns the shape of it's inner tensor.

    Returns:
        Variable (Tensor): The shape of the input variable.

    Examples:
        .. code-block:: python

            >>> import numpy as np
            >>> import paddle
            >>> paddle.enable_static()

            >>> inputs = paddle.static.data(name="x", shape=[3, 100, 100], dtype="float32")
            >>> output = paddle.shape(inputs)

            >>> exe = paddle.static.Executor(paddle.CPUPlace())
            >>> exe.run(paddle.static.default_startup_program())

            >>> img = np.ones((3, 100, 100)).astype(np.float32)

            >>> res = exe.run(paddle.static.default_main_program(), feed={'x':img}, fetch_list=[output])
            >>> print(res)
            [array([  3, 100, 100], dtype=int32)]
    """
    if in_dynamic_or_pir_mode():
        out = _C_ops.shape(input)
        out.stop_gradient = True
        return out
    else:
        check_variable_and_dtype(
            input,
            'input',
            [
                'bool',
                'uint16',
                'float16',
                'float32',
                'float64',
                'int32',
                'int64',
                'complex64',
                'complex128',
                'uint16',
            ],
            'shape',
        )
        helper = LayerHelper('shape', **locals())
        out = helper.create_variable_for_type_inference(dtype='int32')
        helper.append_op(
            type='shape',
            inputs={'Input': input},
            outputs={'Out': out},
            stop_gradient=True,
        )
        out.stop_gradient = True
        return out


def is_complex(x):
    """Return whether x is a tensor of complex data type(complex64 or complex128).

    Args:
        x (Tensor): The input tensor.

    Returns:
        bool: True if the data type of the input is complex data type, otherwise false.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> x = paddle.to_tensor([1 + 2j, 3 + 4j])
            >>> print(paddle.is_complex(x))
            True

            >>> x = paddle.to_tensor([1.1, 1.2])
            >>> print(paddle.is_complex(x))
            False

            >>> x = paddle.to_tensor([1, 2, 3])
            >>> print(paddle.is_complex(x))
            False
    """
    if not isinstance(
        x, (paddle.Tensor, paddle.static.Variable, paddle.pir.Value)
    ):
        raise TypeError(f"Expected Tensor, but received type of x: {type(x)}")
    dtype = x.dtype
    is_complex_dtype = (
        dtype == core.VarDesc.VarType.COMPLEX64
        or dtype == core.VarDesc.VarType.COMPLEX128
        or dtype == core.DataType.COMPLEX64
        or dtype == core.DataType.COMPLEX128
    )
    return is_complex_dtype


def is_floating_point(x):
    """
    Returns whether the dtype of `x` is one of paddle.float64, paddle.float32, paddle.float16, and paddle.bfloat16.

    Args:
        x (Tensor): The input tensor.

    Returns:
        bool: True if the dtype of `x` is floating type, otherwise false.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> x = paddle.arange(1., 5., dtype='float32')
            >>> y = paddle.arange(1, 5, dtype='int32')
            >>> print(paddle.is_floating_point(x))
            True
            >>> print(paddle.is_floating_point(y))
            False
    """
    if not isinstance(
        x, (paddle.Tensor, paddle.static.Variable, paddle.pir.Value)
    ):
        raise TypeError(f"Expected Tensor, but received type of x: {type(x)}")
    dtype = x.dtype
    is_fp_dtype = (
        dtype == core.VarDesc.VarType.FP32
        or dtype == core.VarDesc.VarType.FP64
        or dtype == core.VarDesc.VarType.FP16
        or dtype == core.VarDesc.VarType.BF16
        or dtype == core.DataType.FLOAT32
        or dtype == core.DataType.FLOAT64
        or dtype == core.DataType.FLOAT16
        or dtype == core.DataType.BFLOAT16
    )
    return is_fp_dtype


def is_integer(x):
    """Return whether x is a tensor of integral data type.

    Args:
        x (Tensor): The input tensor.

    Returns:
        bool: True if the data type of the input is integer data type, otherwise false.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> x = paddle.to_tensor([1 + 2j, 3 + 4j])
            >>> print(paddle.is_integer(x))
            False

            >>> x = paddle.to_tensor([1.1, 1.2])
            >>> print(paddle.is_integer(x))
            False

            >>> x = paddle.to_tensor([1, 2, 3])
            >>> print(paddle.is_integer(x))
            True
    """
    if not isinstance(
        x, (paddle.Tensor, paddle.static.Variable, paddle.pir.Value)
    ):
        raise TypeError(f"Expected Tensor, but received type of x: {type(x)}")
    dtype = x.dtype

    is_int_dtype = False
    if not use_pir_api():
        is_int_dtype = (
            dtype == core.VarDesc.VarType.UINT8
            or dtype == core.VarDesc.VarType.INT8
            or dtype == core.VarDesc.VarType.INT16
            or dtype == core.VarDesc.VarType.INT32
            or dtype == core.VarDesc.VarType.INT64
        )
    else:
        is_int_dtype = (
            dtype == core.DataType.UINT8
            or dtype == core.DataType.INT8
            or dtype == core.DataType.INT16
            or dtype == core.DataType.INT32
            or dtype == core.DataType.INT64
        )

    return is_int_dtype


def real(x, name=None):
    """
    Returns a new Tensor containing real values of the input Tensor.

    Args:
        x (Tensor): the input Tensor, its data type could be complex64 or complex128.
        name (str, optional): The default value is None. Normally there is no need for
            user to set this property. For more information, please refer to :ref:`api_guide_Name` .

    Returns:
        Tensor: a Tensor containing real values of the input Tensor.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> x = paddle.to_tensor(
            ...     [[1 + 6j, 2 + 5j, 3 + 4j], [4 + 3j, 5 + 2j, 6 + 1j]])
            >>> print(x)
            Tensor(shape=[2, 3], dtype=complex64, place=Place(cpu), stop_gradient=True,
            [[(1+6j), (2+5j), (3+4j)],
             [(4+3j), (5+2j), (6+1j)]])

            >>> real_res = paddle.real(x)
            >>> print(real_res)
            Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
            [[1., 2., 3.],
             [4., 5., 6.]])

            >>> real_t = x.real()
            >>> print(real_t)
            Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
            [[1., 2., 3.],
             [4., 5., 6.]])
    """
    if in_dynamic_or_pir_mode():
        return _C_ops.real(x)
    else:
        check_variable_and_dtype(x, 'x', ['complex64', 'complex128'], 'real')
        helper = LayerHelper('real', **locals())
        out = helper.create_variable_for_type_inference(
            dtype=_complex_to_real_dtype(helper.input_dtype())
        )
        helper.append_op(type='real', inputs={'X': x}, outputs={'Out': out})
        return out


def imag(x, name=None):
    """
    Returns a new tensor containing imaginary values of input tensor.

    Args:
        x (Tensor): the input tensor, its data type could be complex64 or complex128.
        name (str, optional): The default value is None. Normally there is no need for
            user to set this property. For more information, please refer to :ref:`api_guide_Name` .

    Returns:
        Tensor: a tensor containing imaginary values of the input tensor.

    Examples:
        .. code-block:: python

            >>> import paddle

            >>> x = paddle.to_tensor(
            ...     [[1 + 6j, 2 + 5j, 3 + 4j], [4 + 3j, 5 + 2j, 6 + 1j]])
            >>> print(x)
            Tensor(shape=[2, 3], dtype=complex64, place=Place(cpu), stop_gradient=True,
            [[(1+6j), (2+5j), (3+4j)],
             [(4+3j), (5+2j), (6+1j)]])

            >>> imag_res = paddle.imag(x)
            >>> print(imag_res)
            Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
            [[6., 5., 4.],
             [3., 2., 1.]])

            >>> imag_t = x.imag()
            >>> print(imag_t)
            Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
            [[6., 5., 4.],
             [3., 2., 1.]])
    """
    if in_dynamic_or_pir_mode():
        return _C_ops.imag(x)
    else:
        check_variable_and_dtype(x, 'x', ['complex64', 'complex128'], 'imag')
        helper = LayerHelper('imag', **locals())
        out = helper.create_variable_for_type_inference(
            dtype=_complex_to_real_dtype(helper.input_dtype())
        )
        helper.append_op(type='imag', inputs={'X': x}, outputs={'Out': out})
        return out
